bytecomp/simplif.ml

```diff
@@ -273,15 +278,22 @@ let simplify_exits lam =
       Hashtbl.add subst i ([],simplif l2) ;
       simplif l1
   | Lstaticcatch (l1,(i,xs),l2) ->
-      begin match count_exit i with
-      | 0 -> simplif l1
-      | 1 when i >= 0 ->
-          Hashtbl.add subst i (xs,simplif l2) ;
-          simplif l1
-      | _ ->
-          Lstaticcatch (simplif l1, (i,xs), simplif l2)
-      end
-  | Ltrywith(l1, v, l2) -> Ltrywith(simplif l1, v, simplif l2)
+      let {count; max_depth} = get_exit i in
+      if count = 0 then
+        (* Discard staticcatch: not matching exit *)
+        simplif l1
+      else if i >= 0 && count = 1 && max_depth <= !try_depth then begin
```

Comment:
As a side note, I think you've made the negative index trick irrelevant; you could probably remove the test `i >= 0` and even remove `Lambda.next_negative_raise_count`.
Also, `max_depth < !try_depth` would be very worrying; I'd prefer an `assert false` for this case or at least an equality check instead.

Reply:
> negative index trick irrelevant

Indeed.  I thought it was also used to prevent inlining the handler when the static raise is not in tail context, but apparently this cannot happen.